src/adb.js

```diff
@@ -1 +1,353 @@
+				console.log('start server', err);
+			});
+	}
+
+	/**
+	 * Attempts to find the adb executable, then stop the adb server.
+	 * @returns {Promise}
+	 */
+	stopServer() {
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['kill-server']);
+			})
+			.catch(err => {
+				console.log('stop server', err);
+			});
+	}
+
+	/**
+	 * Forwards the specified device/emulator's socket connections to the destination.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} src - The source port in the format "tcp:<port>"
+	 * @param {String} dest - The destination port in the format "tcp:<port>" or "jdwp:<pid>"
+	 * @returns {Promise}
+	 */
+	forward(deviceId, src, dest) {
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['-s', deviceId, 'forward', src, dest]);
+			})
+			.catch(err => {
+				console.log('stop server', err);
+			});
+	}
+
+	/**
+	 * Pushes a single file to a device or emulator.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} src - The source file to copy to the device
+	 * @param {String} dest - The destination to write the file
+	 * @returns {Promise}
+	 */
+	push(deviceId, src, dest) {
+		src = path.resolve(src);
+		if (!fs.existsSync(src)) {
+			return Promise.reject('Source file "%s" does not exist', src);
+		}
+
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['-s', deviceId, 'push', src, dest]);
+			})
+			.catch(err => {
+				console.log('stop server', err);
+			});
+	}
+
+	/**
+	 * Pulls a single file from a device or emulator.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} src - The source file to copy from the device
+	 * @param {String} dest - The destination to write the file
+	 * @returns {Promise}
+	 */
+	pull(deviceId, src, dest) {
+		dest = path.resolve(dest);
+		const destDir = path.dirname(dest);
+
+		fs.existsSync(destDir) || wrench.mkdirSyncRecursive(destDir);
+
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['-s', deviceId, 'pull', src, dest]);
+			})
+			.catch(err => {
+				console.log('stop server', err);
+			});
+	}
+
+	/**
+	 * Returns the pid of the specified app and device/emulator, if running.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} appid - The application's id
+	 * @returns {Promise}
+	 */
+	getPid(deviceId, appid) {
+		return this.shell(deviceId, 'ps')
+			.then(data => {
+				if (data) {
+					const lines = data.toString().split('\n');
+					for (let i = 0, j = lines.length; i < j; i++ ) {
+						let columns = lines[i].trim().split(/\s+/);
+						if (columns.pop() === appid) {
+							return parseInt(columns[1]);
+						}
+					}
+				}
+				return 0;
+			})
+			.catch(err => {
+				console.log('adb getPid: ', err);
```

Comment:
We don't want to catch errors and we don't want to `console.log()`. This should be done by the consumer of this library.